source/extensions/filters/http/csrf/csrf_filter.cc

```diff
@@ -0,0 +1,114 @@
+#include "extensions/filters/http/csrf/csrf_filter.h"
+
+#include "envoy/stats/scope.h"
+
+#include "common/common/empty_string.h"
+#include "common/http/header_map_impl.h"
+#include "common/http/headers.h"
+#include "common/http/utility.h"
+
+#include "extensions/filters/http/well_known_names.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Csrf {
+
+CsrfFilterConfig::CsrfFilterConfig(const envoy::config::filter::http::csrf::v2::CsrfPolicy& policy,
+                                   const std::string& stats_prefix, Stats::Scope& scope,
+                                   Runtime::Loader& runtime)
+    : stats_(generateStats(stats_prefix, scope)), policy_(generatePolicy(policy, runtime)) {}
+
+CsrfFilter::CsrfFilter(const CsrfFilterConfigSharedPtr config) : config_(config) {}
+
+Http::FilterHeadersStatus CsrfFilter::decodeHeaders(Http::HeaderMap& headers, bool) {
+  determinePolicy();
+
+  if (!policy_->enabled() && !policy_->shadowEnabled()) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  if (!modifyMethod(headers)) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  bool isValid = true;
+  const auto& sourceOrigin = sourceOriginValue(headers);
+  if (sourceOrigin == EMPTY_STRING) {
+    isValid = false;
+    config_->stats().missing_source_origin_.inc();
+  }
+
+  const auto& targetOrigin = targetOriginValue(headers);
+  if (sourceOrigin != targetOrigin) {
+    isValid = false;
+    config_->stats().request_invalid_.inc();
+  }
+
+  if (isValid == true) {
+    config_->stats().request_valid_.inc();
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  if (policy_->shadowEnabled() && !policy_->enabled()) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  callbacks_->sendLocalReply(Http::Code::Forbidden, "Invalid origin", nullptr, absl::nullopt);
+  return Http::FilterHeadersStatus::StopIteration;
+}
+
+bool CsrfFilter::modifyMethod(const Http::HeaderMap& headers) {
+  const auto& method = headers.Method();
+  if (method == nullptr) {
+    return false;
+  }
+  const auto& method_type = method->value().c_str();
+  return (method_type == Http::Headers::get().MethodValues.Post ||
```

Comment:
factor out Http::Headers::get().MethodValues into a temp. In this case it'll have to be an auto since there is no  name for the type :)